provider/core/src/resource.rs

```diff
@@ -74,59 +73,132 @@ impl writeable::Writeable for ResourceCategory {
+    _tag0: [u8; 8],
+    hash: ResourceKeyHash,
+    _tag1: [u8; 2],
 }
 
-/// Shortcut to construct a const resource identifier.
-///
-/// # Examples
-///
-/// Create a private-use ResourceKey:
-///
-/// ```
-/// use icu_provider::prelude::*;
-///
-/// const MY_PRIVATE_USE_KEY: ResourceKey = icu_provider::resource_key!(x, "foo", "bar", 1);
-/// assert_eq!("x-foo/bar@1", format!("{}", MY_PRIVATE_USE_KEY));
-/// ```
-///
-/// Create a ResourceKey for a specific [`ResourceCategory`] (for ICU4X library code only):
-///
-/// ```
-/// use icu_provider::prelude::*;
-///
-/// const MY_PRIVATE_USE_KEY: ResourceKey = icu_provider::resource_key!(Plurals, "ordinal", 1);
-/// assert_eq!("plurals/ordinal@1", format!("{}", MY_PRIVATE_USE_KEY));
-/// ```
-#[macro_export]
-macro_rules! resource_key {
-    ($category:ident, $sub_category:literal, $version:tt) => {
-        $crate::resource_key!($crate::ResourceCategory::$category, $sub_category, $version)
-    };
-    (x, $pu:literal, $sub_category:literal, $version:tt) => {
-        $crate::resource_key!(
-            $crate::ResourceCategory::PrivateUse($crate::internal::tinystr4!($pu)),
-            $sub_category,
-            $version
-        )
-    };
-    ($category:expr, $sub_category:literal, $version:tt) => {
-        $crate::ResourceKey {
-            category: $category,
-            sub_category: $crate::internal::tinystr16!($sub_category),
-            version: $version,
+impl ResourceKey {
+    /// Gets a human-readable representation of a [`ResourceKey`].
+    /// 
+    /// The human-readable path string always contains at least one '/', and it ends with '@'
+    /// followed by one or more digits. Paths do not contain characters other than lowercase
+    /// ASCII, '_', '/', and '@'.
+    /// 
+    /// Useful for reading and writing data to a file system.
+    #[inline]
+    pub const fn get_path(&self) -> &str {
+        self.path
+    }
+
+    /// Gets a machine-readable representation of a [`ResourceKey`].
+    ///
+    /// The machine-readable hash is 4 bytes and can be used as the key in a map.
+    ///
+    /// The hash is a 32-bit FxHash of the path, computed as if on a little-endian platform.
+    #[inline]
+    pub const fn get_hash(&self) -> ResourceKeyHash {
+        self.hash
+    }
+
+    /// Creates a new ResourceKey from a path, panicking if the path is invalid.
+    /// 
+    /// # Panics
+    /// 
+    /// Panics if the syntax of the path is not valid.
+    #[inline]
+    pub const fn new(path: &'static str) -> Self {
+        match Self::try_new(path) {
+            Ok(v) => v,
+            Err(_) => panic!("{}", path) // Invalid syntax!
         }
-    };
+    }
+
+    /// Creates a new ResourceKey from a path, returning an error if the path is invalid.
+    #[inline]
+    pub const fn try_new(path: &'static str) -> Result<Self, ()> {
+        match Self::check_path_syntax(path) {
+            Ok(_) => Ok(Self {
+                path,
+                _tag0: *b"ICU4XK[\x02",
```

Comment:
They may as well be magic numbers.  I have no scientific reason for doing what I did here other than to make it less likely to have collisions.

Reply:
Is a 2 any less likely than a 57? I don't see how any string is less or more likely than any other (maybe other than all `NUL`), so I'd prefer if these were all printable bytes.

If I'm not mistaken the choice of tag length came from the previous PR because it fits exactly into a `tinystr8` and not any architectural or probabilistic arguments. If the bytes in a binary are uniformly distributed then it's very unlikely that 64 consecutive bits match this.

In any case, I'd appreciate removing `_tag1` and making the tag printable (alternatively, use arbitrary numbers).